apiserver/worker/compiler.py

```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
+        remaining_length -= used
+        result.extend(lines)
+        result.append(INSTALL_ERROR_END)
+
+    _, lines = bound_errors(language_detection_errors + compile_errors, remaining_length)
+    result.extend(lines)
+    return result
+
+
+def compile_anything(bot_dir, installTimeLimit=600, timelimit=600, max_error_len=10*1024):
+    install_stdout = []
+    install_errors = []
     if os.path.exists(os.path.join(bot_dir, "install.sh")):
-        _, errors = _run_cmd("chmod +x install.sh; ./install.sh", bot_dir, installTimeLimit)
-    detected_language, errors = detect_language(bot_dir)
+        install_stdout, install_errors = _run_cmd(
+            "chmod +x install.sh; ./install.sh", bot_dir, installTimeLimit)
+
+    detected_language, language_errors = detect_language(bot_dir)
+
     print("detected language " + str(detected_language))
-    if detected_language:
-        print("compiling")
-        compiled, errors = compile_function(detected_language, bot_dir, timelimit)
-        print("done compiling")
-        if compiled:
-            name = detected_language.name
-            run_cmd = detected_language.command
-            run_filename = os.path.join(bot_dir, 'run.sh')
-            print("filename:")
-            print(run_filename)
-            try:
-                with open(run_filename, 'wb') as f:
-                    f.write(('#%s\n%s\n' % (name, run_cmd)).encode('utf-8'))
-                print("file:")
-                with open(run_filename, 'r') as f:
-                    for line in f:
-                        print(line)
-            except Exception as e:
-                print("error")
-                print(e)
-
-            # allow LANGUAGE file to override language name
-            override_name = detect_language_file(bot_dir)
-            if override_name:
-                name = override_name
-            return name, None
-        else:
-            # limit length of reported errors
-            if len(errors) > 0 and sum(map(len, errors)) > max_error_len:
-                first_errors = []
-                cur_error = 0
-                length = len(errors[0])
-                while length < (max_error_len / 3): # take 1/3 from start
-                    first_errors.append(errors[cur_error])
-                    cur_error += 1
-                    length += len(errors[cur_error])
-                first_errors.append("...")
-                length += 3
-                end_errors = []
-                cur_error = -1
-                while length <= max_error_len:
-                    end_errors.append(errors[cur_error])
-                    cur_error -= 1
-                    length += len(errors[cur_error])
-                end_errors.reverse()
-                errors = first_errors + end_errors
-
-            return detected_language.name, errors
-    else:
-        return "Unknown", errors
+    if not detected_language or install_errors:
```

Comment:
I think bailing out here if anything was written to stderr in install.sh is too eager, some programs simply log informational messages to stderr. Currently the compilation is still attempted and only fails if the actual compilation step fails.

Reply:
Maybe checking the exit code from install.sh would work, but I'd be a little worried about even doing that given that we haven't up till now.